src/mpi/comm/comm_split_type.c

```diff
@@ -152,14 +152,52 @@ int MPIR_Comm_split_type_hw_guided(MPIR_Comm * comm_ptr, int key,
     goto fn_exit;
 }
 
-int MPIR_Comm_split_type_hw_unguided(MPIR_Comm * user_comm_ptr, int key,
-                                     MPIR_Info * info_ptr, MPIR_Comm ** newcomm_ptr)
+int MPIR_Comm_split_type_hw_unguided(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcomm_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPIR_Comm *subcomm = NULL;
 
+    int orig_size = MPIR_Comm_size(comm_ptr);
+
+    /* we'll try top-down until we hit upon a subset */
+    mpi_errno = MPIR_Comm_split_type_by_node(comm_ptr, key, &subcomm);
+    MPIR_ERR_CHECK(mpi_errno);
+
```

Comment:
Is it OK that we always assume that the top level resource is node? E.g., maybe we want to start from network-cabinet (a group of nodes) level on a Dragonfly network. Do you think it might be necessary to make code more flexible?

Reply:
We'll have to amend the code to support network topology, to be done in the future. Network topology will just be tried before `split_type_by_node`. So the code is not "inflexible", it just needs to be *extended*.

Reply:
Added `TODO` comment.